packages/core/src/utils/tests/variableValidator.test.ts

```diff
@@ -0,0 +1,47 @@
+import { expect, jest, test } from '@jest/globals';
+import { variableValidator } from '../promptTemplateValidator/variableValidator';
+
+const isPositive = (n) => n > 0;
+const isString = (s) => typeof s === 'string';
+
+let validator;
+validator = new variableValidator(['age', 'name']);
+
+test('add and remove specific rules', () => {
+    expect(() => validator.addSpecificRule('age', isPositive)).not.toThrow();
+
+    expect(() => validator.removeSpecificRule('age')).not.toThrow();
+    expect(() => validator.removeSpecificRule('age')).toThrow();
+});
+
+test('add and remove exclude variables', () => {
+    validator.addExcludeVariable('name');
+    expect(validator.getAttributes().rules.exclude).toContain('name');
+    validator.removeExcludeVariable('name');
+    expect(validator.getAttributes().rules.exclude).not.toContain('name');
+});
+
+test('validate with specific rule', () => {
+    validator.addSpecificRule('age', isPositive);
+    expect(validator.validate({ age: 5 })).toEqual({ isValid: true });
+    expect(validator.validate({ age: -3 })).toEqual({ isValid: false, errorMessage: 'Validation failed for age' });
+});
+
+test('validate with default rule', () => {
+    validator = new variableValidator(['age', 'name'], { default: isString });
+    expect(validator.validate({ age: '25', name: 'John' })).toEqual({ isValid: true });
+    expect(validator.validate({ age: 25, name: 'John' })).toEqual({ isValid: false, errorMessage: 'Validation failed for age' });
+});
+
+test('add specific rule with default rule conflict and exclusion', () => {
+    const validator = new variableValidator(['age', 'name'], { default: isString });
```

Comment:
I really like this implementation design of a validator. It allows the end user to define their rules of checking a certain variable in a JSON object. Nice work!